### PR TITLE
Partly FIFS settlement (netting) and tests (still unstable)

### PR DESCRIPTION
**Pull request by myronrotter**
Devides contracts in `Utility.sol` for basic utility functions and `FifsUtility.sol` for partly FIFS settlement. Refer issue #19.
`FifsUtility.sol` inherrites from `Utility.sol` and overrides functions `addHousehold` and `settle`. 
Further, `FifsUtility.sol` implements first *fair* FIFS netting algorithm in `settle`. 
`fifsUtility.test.js` tests `FifsUtility.sol` for `addHousehold`, `retrieveReward` and for the different cases of `settle`. This PR mainly refers to the second part of issue #11.

Test the contracts: `yarn test-contracts` or `truffle test`.

### CODE REVIEW

File: contracts/FifsUtility.sol [hunk 1]
```diff
@@ -0,0 +1,197 @@
+pragma solidity >=0.5.0 <0.6.0;
+
+import "./Utility.sol";
+
+
+contract FifsUtility is Utility {
+  // number of successful settlements done
+  // used for backtracking
+  uint256 public checkpoint;
```
Comment (dakingha69) [inline contracts/FifsUtility.sol]: Does it make sense to maybe use the current block number for this? Then we could include the additional information at which point the deed was created. Or am I getting something wrong here?

File: contracts/FifsUtility.sol [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    return true;
+  }
+
+  /**
+   * @dev 'Finalizes' an energy transfer by emitting event EnergyTransfer
+   * @param _household address of the household owner/ parity node ?
+   * @param _checkpoint uint256 the settlement number the sender wants to retrieve the reward
+   * @return int256 the total amount of energy transferred by _household in settlement _checkpoint
+   */
+  function retrieveReward(address _household, uint256 _checkpoint)
```
Comment (dakingha69) [inline contracts/FifsUtility.sol]: The naming is a bit misleading IMO. Maybe `proofDeed` or similar would be more appropriate.

File: contracts/FifsUtility.sol [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    Deed[] storage deed = deeds[_household][_checkpoint];
+
+    for (uint256 i = 0; i < deed.length; i++) {
+      require(deed[i].active, "Deed does not exist");
+
+      emit EnergyTransfer(
+        _household,
+        deed[i].to,
+        deed[i].energyTransferred,
+        deed[i].isRenewable);
```
Comment (dakingha69) [inline contracts/FifsUtility.sol]: I think it makes sense to somehow include in here the block number when the deed was created... Maybe use the block number for the `checkpoint`? :thinking:

File: contracts/FifsUtility.sol [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+  // iterable list of all current households with positive amount of renewable energy
+  address[] public householdListWithEnergy;
+  // iterable list of all current households with negative amount of renewable energy
+  address[] public householdListNoEnergy;
+
+  struct Deed {
+    bool active;
+    address to;
+    int256 energyTransferred;
+    bool isRenewable;
```
Comment (dakingha69) [inline contracts/FifsUtility.sol]: from jacob: add `from`

**Comment by myronrotter** [comment]
Fixes #11 and fixes #19

**Pull request merged**